framework/src/utils/SlepcSupport.C

```diff
@@ -526,12 +548,42 @@ moosePetscSNESFormMatricesTags(
   {
     jacobians.emplace_back(libmesh_make_unique<PetscMatrix<Number>>(mat, sys.comm()));
     jacobians.back()->attach_dof_map(sys.get_dof_map());
-    jacobians.back()->zero();
+    if (!eigen_problem->constJacobian())
+      jacobians.back()->zero();
   }
 
   eigen_problem->computeMatricesTags(*sys.current_local_solution.get(), jacobians, tags);
 }
 
+PetscErrorCode
+mooseSlepcEigenFormFunctionMFFD(void * ctx, Vec x, Vec r)
+{
+  PetscErrorCode ierr;
+  PetscErrorCode (*func)(SNES, Vec, Vec, void *);
+  void * fctx;
+  PetscFunctionBegin;
+
+  EigenProblem * eigen_problem = static_cast<EigenProblem *>(ctx);
+  NonlinearEigenSystem & eigen_nl = eigen_problem->getNonlinearEigenSystem();
+  SNES snes = eigen_nl.getSNES();
+
+  eigen_problem->onLinearSolver(true);
+
+  ierr = SNESGetFunction(snes, NULL, &func, &fctx);
+  CHKERRQ(ierr);
+  if (fctx != ctx)
+  {
+    SETERRQ(
+        PetscObjectComm((PetscObject)snes), PETSC_ERR_ARG_INCOMP, "Contexts are not consistent \n");
+  }
```

Comment:
Is there some RAII stuff happening behind the scenes here? I never know what might be happening with macros. If not, then we can remove braces?

Reply:
I will check